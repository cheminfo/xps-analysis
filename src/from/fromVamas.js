import { ensureString } from 'ensure-string';
import { parse } from 'vamas';

import { Analysis } from '..';

import { getNormalizedMeta } from './getNormalizedMeta';

/**
 * Returns an Analysis from a VAMAS text file
 * @param {arrayBuffer|string} [text] the vamas text file
 */
export function fromVamas(text) {
  text = ensureString(text);
  let parsed = parse(text);
  let header = parsed.header;
  let blocks = parsed.blocks;
  let title = header['experiment identifier'];
  let analysis = new Analysis();
  for (let block of blocks) {
    let sourceEnergy = block['analysis source characteristic energy'];

    let yVariable = block.correspondingVariables[0];
    let yLabel = yVariable.label;
    let yValues = yVariable.array;

    let xLabel = block['abscissa label'];
    let xUnits = block['abscissa units'];
    let xStart = block['abscissa start'];
    let xIncrement = block['abscissa increment'];
    let xValues = new Float64Array(yValues.length);
    for (let i = 0; i < yValues.length; i++) {
      xValues[i] = xStart + xIncrement * i;
    }
    // currently we take the first corresponding variables

    let meta = {};
    for (let key in block) {
      if (typeof block[key] === 'string' || typeof block[key] === 'number') {
        meta[key] = block[key];
      }
    }

    for (let key in header) {
      if (typeof header[key] === 'string' || typeof header[key] === 'number') {
        meta[key] = header[key];
      }
    }

    meta.cheminfo = { meta: getNormalizedMeta(meta) };

    const variables = {};
    if (xLabel === 'Kinetic energy' && sourceEnergy) {
      // we will calculate binding energy
      variables.x = {
        data: xValues.map((value) => sourceEnergy - value).reverse(),
        label: 'Binding energy',
        units: xUnits,
        type: 'DEPENDENT',
      };
      variables.y = {
        data: yValues.reverse(),
        label: yLabel,
        type: 'DEPENDENT',
      };
      variables.k = {
        data: xValues.reverse(),
        label: xLabel,
        units: xUnits,
        type: 'INDEPENDENT',
      };
    } else {
      variables.x = {
        data: xValues,
        label: xLabel,
        units: xUnits,
        type: 'INDEPENDENT',
      };
      variables.y = {
        data: yValues,
        label: yLabel,
        type: 'DEPENDENT',
      };
    }

    analysis.pushSpectrum(variables, {
      dataType: 'XPS',
      title,
      meta,
    });
  }

  return analysis;
}
