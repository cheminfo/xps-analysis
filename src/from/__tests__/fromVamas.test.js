import { readFileSync } from 'fs';
import { join } from 'path';

import { fromVamas } from '../fromVamas';

describe('fromVamas', () => {
  const text = readFileSync(
    join(__dirname, '../../../testFiles/multiplex.vms'),
    'utf8',
  );

  it('multiplex', () => {
    let result = fromVamas(text).spectra[0];
    expect(Object.keys(result.variables)).toHaveLength(3);
    expect(Object.keys(result.meta)).toHaveLength(61);
  });
});
