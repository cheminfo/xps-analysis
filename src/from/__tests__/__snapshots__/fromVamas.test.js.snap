// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fromVamas Cellulose with CASA information, regions and components 1`] = `
{
  "analyserMode": "FAT",
  "analysisSource": {
    "beamWidthX": {
      "units": "um",
      "value": 6000,
    },
    "beamWidthY": {
      "units": "um",
      "value": 500,
    },
    "characteristicEnergy": {
      "units": "eV",
      "value": 1486.6,
    },
    "label": "Al Ka, Monochromated",
  },
  "components": [
    {
      "area": {
        "lowerBound": 1e-20,
        "upperBound": 126904.8,
        "value": 4830.0138,
      },
      "assignment": "C 1s 1",
      "atomicMass": 12.011,
      "bindingEnergy": {
        "units": "eV",
        "value": 286.7736,
      },
      "componentID": "C 1s 1",
      "fwhm": {
        "lowerBound": 0.25,
        "upperBound": 6.25,
        "value": 1.1057933,
      },
      "groupIndex": -1,
      "kineticEnergy": {
        "units": "eV",
        "value": 1199.8264,
      },
      "position": {
        "lowerBound": 1199.7,
        "upperBound": 1199.87,
        "value": 1199.8264,
      },
      "relativeSensitivityFactor": 0.278,
      "shape": {
        "kind": "lorentzianAsymmetric",
        "options": {
          "alpha": 1.4,
          "beta": 1.7,
          "m": 100,
        },
      },
      "uncorrectedRSF": 0.278,
    },
    {
      "area": {
        "constrain": {
          "factor": 0.2048,
          "linkedComponent": 0,
        },
        "lowerBound": 0,
        "upperBound": 2000,
        "value": 989.18682,
      },
      "assignment": "C 1s 2",
      "atomicMass": 12.011,
      "bindingEnergy": {
        "units": "eV",
        "value": 288.1950999999999,
      },
      "componentID": "C 1s 2",
      "fwhm": {
        "lowerBound": 0.25,
        "upperBound": 6.25,
        "value": 1.1121055,
      },
      "groupIndex": -1,
      "kineticEnergy": {
        "units": "eV",
        "value": 1198.4049,
      },
      "position": {
        "lowerBound": 1198.34,
        "upperBound": 1198.74,
        "value": 1198.4049,
      },
      "relativeSensitivityFactor": 0.278,
      "shape": {
        "kind": "lorentzianAsymmetric",
        "options": {
          "alpha": 1.4,
          "beta": 1.6,
          "m": 100,
        },
      },
      "uncorrectedRSF": 0.278,
    },
    {
      "area": {
        "lowerBound": 0,
        "upperBound": 243,
        "value": 56.812945,
      },
      "assignment": "C 1s degrad",
      "atomicMass": 12.011,
      "bindingEnergy": {
        "units": "eV",
        "value": 288.8453999999999,
      },
      "componentID": "C 1s degrad",
      "fwhm": {
        "lowerBound": 0.25,
        "upperBound": 2,
        "value": 0.82920074,
      },
      "groupIndex": -1,
      "kineticEnergy": {
        "units": "eV",
        "value": 1197.7546,
      },
      "position": {
        "lowerBound": 1197.3546,
        "upperBound": 1197.7546,
        "value": 1197.7546,
      },
      "relativeSensitivityFactor": 0.278,
      "shape": {
        "kind": "gaussianLorentzianProduct",
        "options": {
          "mixingRatio": 30,
        },
      },
      "uncorrectedRSF": 0.278,
    },
    {
      "area": {
        "lowerBound": 1e-20,
        "upperBound": 126904.8,
        "value": 100.40835,
      },
      "assignment": "C 1s Adv.C",
      "atomicMass": 12.011,
      "bindingEnergy": {
        "units": "eV",
        "value": 284.8711999999998,
      },
      "componentID": "C 1s Adv.C",
      "fwhm": {
        "lowerBound": 0.25,
        "upperBound": 6.25,
        "value": 0.74742285,
      },
      "groupIndex": -1,
      "kineticEnergy": {
        "units": "eV",
        "value": 1201.7288,
      },
      "position": {
        "lowerBound": 1201.4,
        "upperBound": 1201.8,
        "value": 1201.7288,
      },
      "relativeSensitivityFactor": 0.278,
      "shape": {
        "kind": "gaussianLorentzianProduct",
        "options": {
          "mixingRatio": 30,
        },
      },
      "uncorrectedRSF": 0.278,
    },
  ],
  "energyType": {
    "kind": "binding",
    "units": "eV",
  },
  "from": 294.16,
  "region": {
    "auger": {
      "element": null,
      "transition": null,
    },
    "name": "C 1s",
    "orbital": {
      "angularMomentum": "s",
      "element": "C",
      "shell": 1,
    },
  },
  "regions": [
    {
      "background": {
        "parameters": {
          "averageWidth": 1,
          "bindingEnergyEnd": {
            "units": "eV",
            "value": 283.34839999999986,
          },
          "bindingEnergyStart": {
            "units": "eV",
            "value": 293.2871,
          },
          "crossSection": [
            40.442441,
            -450,
            0,
            0,
          ],
          "endOffset": 0,
          "kineticEnergyEnd": {
            "units": "eV",
            "value": 1203.2516,
          },
          "kineticEnergyStart": {
            "units": "eV",
            "value": 1193.3129,
          },
          "relativeSensitivityFactor": 0.278,
          "startOffset": 0,
        },
        "rawParameters": "1193.3129 1203.2516 0.278 2 0 0 40.442441 -450 0 0",
        "type": "Shirley",
      },
      "block": {
        "atomicMass": 12.011,
        "regionBlockID": "C 1s",
        "relativeSensitivityFactor": 0.278,
      },
      "regionID": "C 1s",
    },
  ],
  "speciesLabel": "C",
  "to": 282.16,
}
`;
